File: modules/parallelize.py

import os
import math
import signal
from multiprocessing import Pool
import multiprocessing as mp
from time import time


from modules import help_functions
from modules import cluster


def reads_to_clusters_helper(arguments):
    for k,v in arguments.items():
        args, kwargs = v
    return cluster.reads_to_clusters(*args, **kwargs)


def merge_dicts(*dict_args):
    """
    Given any number of dicts, shallow copy and merge into a new dict,
    precedence goes to key value pairs in latter dicts.
    """
    result = {}
    for dictionary in dict_args:
        result.update(dictionary)
    return result


def batch_list(lst, nr_cores=1, batch_type = "nr_reads" , merge_consecutive = False ):
    if merge_consecutive:
        batch_id = 2
        batch = []
        for info in lst:
            if info[1] <= batch_id:
                batch.append(info)
            else: # first sequence in new batch
                yield batch
                batch_id += 2
                batch = []
                batch.append(info)
        yield batch

    else:
        if batch_type == "nr_reads":
            l = len(lst)
            chunk_size = int(l/nr_cores) + 1
            for ndx in range(0, l, chunk_size):
                yield lst[ndx:min(ndx + chunk_size, l)]
        
        elif batch_type == "total_nt":
            tot_length = sum([len(seq) for i, b_i, acc, seq, qual, score in lst] )
            nt_chunk_size = int(tot_length/nr_cores) + 1

            batch = []
            curr_size = 0
            for info in lst:
                curr_size += len(info[3])
                batch.append(info)
                if curr_size >= nt_chunk_size:
                    yield batch
                    batch = []
                    curr_size = 0
            yield batch   
        
        elif batch_type == "read_lengths_squared":
            tot_length = sum([ math.pow(len(seq),2) for i, b_i, acc, seq, qual, score in lst] )
            nt_chunk_size = int(tot_length/nr_cores) + 1
            batch = []
            curr_size = 0
            for info in lst:
                curr_size +=  math.pow(len(info[3]),2) 
                batch.append(info)
                if curr_size >= nt_chunk_size:
                    yield batch
                    batch = []
                    curr_size = 0
            yield batch               



def print_intermediate_results(clusters, cluster_seq_origin, args, iter_nr):
    path = args.outfolder +"/{0}".format(iter_nr) 
    help_functions.mkdir_p( path )
    outfile = open(os.path.join(path, "pre_clusters.csv"), "w")
    nontrivial_cluster_index = 0
    for c_id, all_read_acc in sorted(clusters.items(), key = lambda x: len(x[1]), reverse=True):
        for r_acc in all_read_acc:
            outfile.write("{0}\t{1}\n".format(c_id, "_".join([item for item in r_acc.split("_")[:-1]]) ))
        if len(all_read_acc) > 1:
            nontrivial_cluster_index += 1
    print("Nr clusters larger than 1:", nontrivial_cluster_index) #, "Non-clustered reads:", len(archived_reads))
    print("Nr clusters (all):", len(clusters)) #, "Non-clustered reads:", len(archived_reads))


    origins_outfile = open(os.path.join(path,  "cluster_origins.csv"), "w")
    for cl_id, all_read_acc in sorted(clusters.items(), key = lambda x: len(x[1]), reverse=True):
        read_cl_id, b_i, acc, c_seq, c_qual, score, error_rate = cluster_seq_origin[cl_id]
        origins_outfile.write("{0}\t{1}\t{2}\t{3}\t{4}\t{5}\n".format(read_cl_id, acc, c_seq, c_qual, score, error_rate)) 
    outfile.close()
    origins_outfile.close()


def parallel_clustering(read_array, p_emp_probs, args):
    num_batches = args.nr_cores 
    read_batches = [batch for batch in batch_list(read_array, num_batches, batch_type = args.batch_type )]
    print("Using total nucleotide batch sizes:", [sum([len(seq) for i, b_i, acc, seq, qual, score in b]) for b in read_batches] )
    print("Nr reads in batches:", [len(b)  for b in read_batches] )
    cluster_batches = []
    cluster_seq_origin_batches = []
    lowest_batch_index_db = []
    for batch in read_batches:
        tmp_clust = {}
        tmp_clust_origin = {}
        for i, b_i, acc, seq, qual, score in batch:
            tmp_clust[i] = [acc]
            tmp_clust_origin[i] = (i, b_i, acc, seq, qual, score)
        cluster_batches.append(tmp_clust)
        cluster_seq_origin_batches.append(tmp_clust_origin)
        lowest_batch_index_db.append({})
    del read_array

    ####### parallelize alignment #########
    # pool = Pool(processes=mp.cpu_count())
    original_sigint_handler = signal.signal(signal.SIGINT, signal.SIG_IGN)
    signal.signal(signal.SIGINT, original_sigint_handler)
    try:
        mp.set_start_method('spawn')
        print("Environment set:", mp.get_context())
    except RuntimeError:
        print("Environment already set:", mp.get_context())
    it = 1
    while True:
        # Structure up batches
        print()
        print("ITERATION", it)
        print("Using {0} batches.".format(num_batches))

        if len(read_batches) == 1:
            start_cluster = time()

            data = {i+1 :((cluster_batches[0], cluster_seq_origin_batches[0], read_batches[0], p_emp_probs, lowest_batch_index_db[0], 1, args), {})} 
            result = reads_to_clusters_helper(data) # { new_batch_index : (Cluster, cluster_seq_origin, H, new_batch_index)}
            Cluster, cluster_seq_origin, _, _ = result[1]
            print("Time elapesd clustering last iteration single core:", time() - start_cluster)
            return Cluster, cluster_seq_origin


        start_multi = time()
        pool = Pool(processes=int(num_batches))
        try:
            # print([len(b) for b in read_batches])
            data = [ {i+1 :((cluster_batches[i], cluster_seq_origin_batches[i], read_batches[i], p_emp_probs, lowest_batch_index_db[i], i+1, args), {})} for i in range(len(read_batches))]
            res = pool.map_async(reads_to_clusters_helper, data)
            cluster_results =res.get(999999999) # Without the timeout this blocking call ignores all signals.
        except KeyboardInterrupt:
            print("Caught KeyboardInterrupt, terminating workers")
            pool.terminate()
            sys.exit()
        else:
            pool.close()
        pool.join()

        print("Time elapesd multiprocessing:", time() - start_multi)

        start_joining = time()
        all_repr = [] # all_repr = [top_new_seq_origins]
        all_cl = []
        all_minimizer_databases = {}
        for output_dict in cluster_results:
            print("New batch")
            for k, v in output_dict.items():
                new_clusters, new_representatives, minimizer_database_new, batch_index = v
                print("Batch index", k)
            # for new_clusters, new_representatives, minimizer_database_new, batch_index in cluster_results: 
                all_cl.append(new_clusters)
                all_repr.append(new_representatives)
                all_minimizer_databases[batch_index] =  minimizer_database_new

        all_clusters = merge_dicts(*all_cl)
        all_representatives = merge_dicts(*all_repr)
        read_array =  [ (i, b_index, acc, seq, qual, score) for i, (i, b_index, acc, seq, qual, score, error_rate) in sorted(all_representatives.items(), key=lambda x: x[1][5], reverse=True)] 
        new_nr_repr = len(read_array)
        print("number of representatives left to cluster:", new_nr_repr)
        print("Time elapesd joining clusters:", time() - start_joining)

        # Determine new number of batches
        if num_batches == 1:
            return all_clusters, all_representatives
        else:
            print_intermediate_results(all_clusters, all_representatives, args, it)

        it += 1
        read_batches = [batch for batch in batch_list(read_array, num_batches, batch_type = args.batch_type, merge_consecutive = True)]
        num_batches = len(read_batches)
        print("Batches after pairwise consecutive merge:", num_batches)
        print("Using total nucleotide batch sizes:", [sum([len(seq) for i, b_i, acc, seq, qual, score in b]) for b in read_batches] )
        print("Using nr reads batch sizes:", [len(b)  for b in read_batches] )
        cluster_batches = []
        cluster_seq_origin_batches = []
        lowest_batch_index_db = []
        for batch in read_batches:
            tmp_clust = {}
            tmp_clust_origin = {}
            lowest_batch_index = min( [ prev_batch_index for (read_cl_id, prev_batch_index, acc, seq, qual, score) in batch ] )
            for i, b_i, acc, seq, qual, score in batch:
                tmp_clust[i] = all_clusters[i]
                tmp_clust_origin[i] = all_representatives[i]
            cluster_batches.append(tmp_clust)
            cluster_seq_origin_batches.append(tmp_clust_origin)
            lowest_batch_index_db.append( all_minimizer_databases[lowest_batch_index])

        del all_minimizer_databases

File: rust/src/parallelize.rs
//! `modules/parallelize.py` -- the `--t > 1` path.
//!
//! **This is a different algorithm, not a parallelised version of the same one.**
//! Reads are cut into batches, each batch is clustered independently, the
//! surviving representatives are pooled and re-batched, and the whole thing
//! repeats until one batch remains. `--t` therefore changes the answer: on the
//! smoke corpus `--t 1` gives 40 clusters and `--t 8` gives 35. See PORTING.md,
//! Finding 3.
//!
//! **The batches are independent, so execution order is free.** Each worker
//! receives its own cluster map, representative map, read slice and minimizer
//! database, and returns new ones; nothing is shared. `pool.map_async` preserves
//! result order, so running the batches sequentially gives byte-identical output
//! to running them in parallel. That is why this is written sequentially first:
//! the batching structure is what has to be exact, and threading it afterwards
//! is behaviour-neutral.

use crate::sweep::{self, OrderedClusters, ReadInfo, SweepParams, SweepRead};
use rustc_hash::FxHashMap;
use std::collections::HashMap;

/// `batch_list`'s three real modes.
///
/// Note the CLI's help advertises `"weighted"`, which **no branch implements** --
/// the third mode is spelled `read_lengths_squared`. Passing the documented
/// value yields no batches at all and the reference then dies with
/// `ValueError: Number of processes must be at least 1`. PORTING.md, Finding 15.
#[derive(Clone, Copy, PartialEq)]
pub enum BatchType {
    NrReads,
    TotalNt,
    ReadLengthsSquared,
}

impl BatchType {
    pub fn parse(s: &str) -> Option<BatchType> {
        match s {
            "nr_reads" => Some(BatchType::NrReads),
            "total_nt" => Some(BatchType::TotalNt),
            "read_lengths_squared" => Some(BatchType::ReadLengthsSquared),
            _ => None,
        }
    }
}

/// `batch_list(lst, nr_cores, batch_type)` -- the first-pass split.
pub fn batch_list(reads: &[SweepRead], nr_cores: usize, bt: BatchType) -> Vec<Vec<SweepRead>> {
    let mut out: Vec<Vec<SweepRead>> = Vec::new();
    match bt {
        BatchType::NrReads => {
            // `chunk_size = int(l / nr_cores) + 1`, so the last chunk can be
            // short and there can be fewer chunks than cores.
            let l = reads.len();
            let chunk = l / nr_cores.max(1) + 1;
            let mut ndx = 0usize;
            while ndx < l {
                out.push(reads[ndx..(ndx + chunk).min(l)].to_vec());
                ndx += chunk;
            }
        }
        BatchType::TotalNt | BatchType::ReadLengthsSquared => {
            let weight = |r: &SweepRead| -> f64 {
                let n = r.seq.len() as f64;
                if bt == BatchType::TotalNt {
                    n
                } else {
                    // math.pow(len, 2)
                    n * n
                }
            };
            // `tot_length` is a sum of ints for total_nt and of floats for the
            // squared variant; `int(tot/nr_cores) + 1` truncates either way.
            let total: f64 = reads.iter().map(weight).sum();
            let chunk = (total / nr_cores.max(1) as f64) as i64 + 1;
            let mut batch: Vec<SweepRead> = Vec::new();
            let mut curr = 0f64;
            for r in reads {
                curr += weight(r);
                batch.push(r.clone());
                if curr >= chunk as f64 {
                    out.push(std::mem::take(&mut batch));
                    curr = 0.0;
                }
            }
            // UNCONDITIONAL, and that is the difference from isONclust.
            //
            // isONclust guards all three of its final yields with `if batch:`;
            // NGSpeciesID does not, so when the last chunk lands exactly on the
            // threshold it yields an EMPTY batch. That is not cosmetic: the
            // batch count drives the merge-iteration count, which decides how
            // many numbered `<n>/` directories a run writes -- and those are
            // recorded output.
            //
            // The two repositories fixed the same underlying defect in two
            // different places. isONclust stopped yielding empty batches;
            // NGSpeciesID instead wrote `min(prev_b_indices or [1])` in
            // cluster.py, which covers ONE of the two `min()` call sites. The
            // other is parallelize.py's own, and it is why `--batch_type
            // weighted` crashes there (Finding 5).
            out.push(batch);
        }
    }
    out
}

/// `batch_list(..., merge_consecutive=True)` -- the re-batching between rounds.
///
/// This walks the score-ordered representatives and closes a batch whenever a
/// read's *previous batch index* exceeds a threshold that starts at 2 and rises
/// by 2 each time. The reads are in score order, not batch order, so the group
/// boundaries fall wherever the batch indices happen to cross the threshold --
/// arbitrary, but deterministic, and it is what decides which minimizer database
/// each group inherits.
pub fn batch_list_merge_consecutive(reads: &[SweepRead]) -> Vec<Vec<SweepRead>> {
    let mut out: Vec<Vec<SweepRead>> = Vec::new();
    let mut batch_id: i64 = 2;
    let mut batch: Vec<SweepRead> = Vec::new();
    for r in reads {
        if r.prev_batch_index <= batch_id {
            batch.push(r.clone());
        } else {
            // The reference yields the batch even when it is empty.
            out.push(std::mem::take(&mut batch));
            batch_id += 2;
            batch.push(r.clone());
        }
    }
    // Unconditional here too; see `batch_list`.
    out.push(batch);
    out
}

/// What a completed round hands back.
pub struct ParallelResult {
    pub clusters: OrderedClusters,
    pub representatives: FxHashMap<usize, ReadInfo>,
    /// The first failure while writing a merge iteration's two files, already
    /// formatted for stderr. The files are written as each iteration finishes
    /// rather than collected and handed back, so this is how the failure gets
    /// out; the caller reports it and exits non-zero, as it did when the writing
    /// lived there.
    pub intermediate_error: Option<String>,
    pub mapped_passed: usize,
    pub aln_passed: usize,
    pub aln_called: usize,
    /// Reads whose homopolymer-compressed length was under k. The reference
    /// prints one line per read to stdout; a count is more useful and stdout is
    /// not in the byte-identity contract.
    pub skipped_short: usize,
    pub times: sweep::StageTimes,
}

/// `parallel_clustering`.
pub fn parallel_clustering(
    read_array: &[SweepRead],
    nr_cores: usize,
    bt: BatchType,
    table: &crate::p_emp::Table,
    p: SweepParams,
    // `sorted.fastq`, re-read for the representatives' quality strings when an
    // intermediate is written; the clustering stage does not keep them. See
    // `crate::record_ranges_for`.
    sorted_path: &std::path::Path,
    // Where the numbered per-iteration directories go. The writing used to
    // happen in the caller, after this function returned, which meant every
    // iteration's two files stayed resident until then.
    outfolder: &std::path::Path,
) -> ParallelResult {
    let mut num_batches = nr_cores;
    let mut read_batches = batch_list(read_array, num_batches, bt);

    // Per batch: its own cluster map, representative map and (empty) database.
    let mut cluster_batches: Vec<OrderedClusters> = Vec::new();
    let mut rep_batches: Vec<FxHashMap<usize, ReadInfo>> = Vec::new();
    let mut db_batches: Vec<crate::cluster::MinimizerDatabase> = Vec::new();
    for batch in &read_batches {
        let mut c = OrderedClusters::default();
        let mut r = FxHashMap::default();
        for x in batch {
            c.insert(x.id, vec![x.id as u32]);
            r.insert(
                x.id,
                ReadInfo {
                    id: x.id,
                    batch_index: x.prev_batch_index,
                    acc: x.acc.clone(),
                    seq: x.seq.clone(),
                    err_per_base: x.err_per_base,
                    score: x.score,
                    error_rate: None,
                    compressed_len: None,
                },
            );
        }
        cluster_batches.push(c);
        rep_batches.push(r);
        db_batches.push(crate::cluster::MinimizerDatabase::new());
    }

    let mut out = ParallelResult {
        clusters: OrderedClusters::default(),
        representatives: FxHashMap::default(),
        intermediate_error: None,
        mapped_passed: 0,
        aln_passed: 0,
        aln_called: 0,
        skipped_short: 0,
        times: sweep::StageTimes::default(),
    };

    let mut it = 1usize;
    loop {
        // The one-batch case returns straight out of the loop, without a merge
        // iteration and without writing intermediates.
        if read_batches.len() == 1 {
            let res = sweep::reads_to_clusters(
                std::mem::take(&mut cluster_batches[0]),
                std::mem::take(&mut rep_batches[0]),
                &read_batches[0],
                std::mem::take(&mut db_batches[0]),
                1,
                table,
                p,
            );
            out.mapped_passed += res.mapped_passed;
            out.aln_passed += res.aln_passed;
            out.aln_called += res.aln_called;
            out.skipped_short += res.skipped_short;
            out.times.add(&res.times);
            out.clusters = res.clusters;
            out.representatives = res.representatives;
            return out;
        }

        // The pool. One thread per batch, matching the reference's
        // `Pool(processes=num_batches)`.
        //
        // This is behaviour-neutral, and the reason is worth stating rather than
        // hoping: each worker owns its cluster map, representative map, read
        // slice and minimizer database, and returns new ones -- nothing is
        // shared and nothing is mutated across batches. `map_async` preserves
        // result order and so does joining the handles in order, so the merge
        // below sees the same sequence whatever order the threads finish in.
        // Verified, not assumed: all 27 equivalence cases pass either way.
        let inputs: Vec<_> = (0..read_batches.len())
            .map(|i| {
                (
                    std::mem::take(&mut cluster_batches[i]),
                    std::mem::take(&mut rep_batches[i]),
                    &read_batches[i],
                    std::mem::take(&mut db_batches[i]),
                    i as i64 + 1,
                )
            })
            .collect();
        let results: Vec<sweep::SweepResult> = std::thread::scope(|scope| {
            let handles: Vec<_> = inputs
                .into_iter()
                .map(|(c, r, reads, db, idx)| {
                    scope.spawn(move || sweep::reads_to_clusters(c, r, reads, db, idx, table, p))
                })
                .collect();
            handles
                .into_iter()
                .map(|h| h.join().expect("a clustering worker panicked"))
                .collect()
        });
        for res in &results {
            out.mapped_passed += res.mapped_passed;
            out.aln_passed += res.aln_passed;
            out.aln_called += res.aln_called;
            out.skipped_short += res.skipped_short;
            out.times.add(&res.times);
        }

        // merge_dicts: later dicts win, but the batches are disjoint by read id.
        let mut all_clusters = OrderedClusters::default();
        let mut all_reps: FxHashMap<usize, ReadInfo> = FxHashMap::default();
        let mut dbs: HashMap<i64, crate::cluster::MinimizerDatabase> = HashMap::new();
        for res in results {
            // `into_ordered`, not `iter()` + `clone()`: the batches are disjoint
            // by read id and `res` is dead after this, so there is no reason for
            // both copies of every member list to be live at once.
            for (id, members) in res.clusters.into_ordered() {
                all_clusters.insert(id, members);
            }
            all_reps.extend(res.representatives);
            dbs.insert(res.batch_index, res.db);
        }

        // The survivors, re-sorted by score descending.
        let mut survivors: Vec<SweepRead> = all_reps
            .values()
            .map(|r| SweepRead {
                id: r.id,
                prev_batch_index: r.batch_index,
                acc: r.acc.clone(),
                seq: r.seq.clone(),
                hp_error_rate: r.error_rate,
                err_per_base: r.err_per_base,
                score: r.score,
            })
            .collect();
        // `sorted(all_representatives.items(), key=lambda x: x[1][5], reverse=True)`
        // -- a dict of ints, so its own order is deterministic but not insertion
        // order. Ties are broken by id here to keep this reproducible; see the
        // note in PORTING.md.
        survivors.sort_by(|a, b| {
            b.score
                .partial_cmp(&a.score)
                .expect("scores are finite")
                .then(a.id.cmp(&b.id))
        });

        if num_batches == 1 {
            out.clusters = all_clusters;
            out.representatives = all_reps;
            return out;
        }
        // Written now, not collected: `it` is the directory number, and the
        // first failure is carried out in `intermediate_error` for the caller to
        // report. A later iteration is still attempted, which is what happened
        // when the caller did all the writing after the fact.
        if out.intermediate_error.is_none() {
            if let Err(e) = write_intermediate(
                &outfolder.join(it.to_string()),
                &all_clusters,
                &all_reps,
                read_array,
                sorted_path,
            ) {
                out.intermediate_error = Some(e);
            }
        }

        it += 1;
        read_batches = batch_list_merge_consecutive(&survivors);
        num_batches = read_batches.len();
        if num_batches == 0 {
            // The reference reaches Pool(processes=0) and dies with
            // "ValueError: Number of processes must be at least 1".
            out.clusters = all_clusters;
            out.representatives = all_reps;
            return out;
        }

        cluster_batches.clear();
        rep_batches.clear();
        db_batches.clear();
        for batch in &read_batches {
            let mut c = OrderedClusters::default();
            let mut r = FxHashMap::default();
            let lowest = batch
                .iter()
                .map(|x| x.prev_batch_index)
                .min()
                .expect("batches are non-empty here");
            for x in batch {
                if let Some(v) = all_clusters.map.get(&x.id) {
                    c.insert(x.id, v.clone());
                }
                if let Some(v) = all_reps.get(&x.id) {
                    r.insert(x.id, v.clone());
                }
            }
            cluster_batches.push(c);
            rep_batches.push(r);
            db_batches.push(dbs.remove(&lowest).unwrap_or_default());
        }
    }
}

/// `print_intermediate_results` -- the two files written per merge iteration.
///
/// Both sort by cluster size only, with ties falling through to insertion order.
/// `pre_clusters.csv` uses the **internal** cluster id, not a renumbered one, and
/// `cluster_origins.csv` writes the accession **with** its score suffix -- unlike
/// `final_cluster_origins.tsv`, which strips it.
///
/// Streamed, and that is the whole point of it. This used to render both files
/// into `String`s which `parallel_clustering` then held until it returned: on
/// 100k Drosophila reads with `--t 8` that is 137 MB resident for the entire
/// run, growing with reads x read length, and every byte of it was already
/// destined for a file. The representatives' quality strings are no longer
/// slurped either -- each is needed exactly once, so a 12-byte byte range
/// replaces a ~1.4 KB string. Both files are byte-for-byte what they were; all
/// six `--t > 1` equivalence cases pin them.
///
/// An IO failure reading `sorted.fastq` back leaves the quality column empty
/// rather than failing the run, which is what the slurping version did with its
/// `unwrap_or_default`. Preserved deliberately: this is a diagnostic file, and
/// the clustering itself is unaffected.
fn write_intermediate(
    dir: &std::path::Path,
    clusters: &OrderedClusters,
    reps: &FxHashMap<usize, ReadInfo>,
    reads: &[SweepRead],
    sorted_path: &std::path::Path,
) -> Result<(), String> {
    use std::io::Write;
    // The two messages are the caller's former ones, unchanged.
    let wr = |e: std::io::Error| format!("isONclust: cannot write intermediates: {e}");
    std::fs::create_dir_all(dir)
        .map_err(|e| format!("isONclust: cannot create {}: {e}", dir.display()))?;

    let mut order: Vec<usize> = clusters.order.clone();
    order.sort_by(|a, b| clusters.map[b].len().cmp(&clusters.map[a].len()));

    let mut w = std::io::BufWriter::with_capacity(
        1 << 16,
        std::fs::File::create(dir.join("pre_clusters.csv")).map_err(wr)?,
    );
    for c_id in &order {
        for id in &clusters.map[c_id] {
            writeln!(
                w,
                "{}\t{}",
                c_id,
                crate::pipeline::strip_score(&reads[*id as usize].acc)
            )
            .map_err(wr)?;
        }
    }
    // BufWriter swallows the error from its own drop, so the flush is explicit.
    w.flush().map_err(wr)?;

    // Byte ranges for this pass's representatives, and one file handle to read
    // them back through. `sorted.fastq` was written by this process, so these
    // are page-cache hits.
    let want: rustc_hash::FxHashSet<usize> = reps.keys().copied().collect();
    let ranges = crate::pipeline::record_ranges_for(sorted_path, &want).unwrap_or_default();
    let src = std::fs::File::open(sorted_path).ok();

    let mut w = std::io::BufWriter::with_capacity(
        1 << 16,
        std::fs::File::create(dir.join("cluster_origins.csv")).map_err(wr)?,
    );
    let mut seq = Vec::new();
    let mut raw = Vec::new();
    for c_id in &order {
        let r = &reps[c_id];
        r.seq.unpack_into(&mut seq);
        let qual = match (&src, ranges.get(c_id)) {
            (Some(f), Some(&(at, n))) => crate::pipeline::qual_at(f, at, n, &mut raw)
                .ok()
                .flatten()
                .unwrap_or_default(),
            _ => String::new(),
        };
        writeln!(
            w,
            "{}\t{}\t{}\t{}\t{}\t{}",
            r.id,
            r.acc,
            String::from_utf8_lossy(&seq),
            qual,
            crate::pyfloat::repr(r.score),
            crate::pyfloat::repr(r.error_rate.unwrap_or(f64::NAN)),
        )
        .map_err(wr)?;
    }
    w.flush().map_err(wr)?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn r(id: usize, b: i64, len: usize, score: f64) -> SweepRead {
        SweepRead {
            id,
            prev_batch_index: b,
            acc: format!("r{id}_{score}").into(),
            seq: crate::packed::PackedSeq::from_bytes(&vec![b'A'; len]).0,
            hp_error_rate: None,
            err_per_base: 0.0,
            score,
        }
    }

    #[test]
    fn batch_type_weighted_is_not_implemented() {
        // The CLI documents it; no branch handles it. Finding 15.
        assert!(BatchType::parse("weighted").is_none());
        assert!(BatchType::parse("total_nt").is_some());
        assert!(BatchType::parse("nr_reads").is_some());
        assert!(BatchType::parse("read_lengths_squared").is_some());
    }

    #[test]
    fn nr_reads_uses_chunk_size_len_over_cores_plus_one() {
        let reads: Vec<SweepRead> = (0..10).map(|i| r(i, 0, 100, 1.0)).collect();
        // 10/4 + 1 = 3, so 3+3+3+1
        let b = batch_list(&reads, 4, BatchType::NrReads);
        assert_eq!(
            b.iter().map(|x| x.len()).collect::<Vec<_>>(),
            vec![3, 3, 3, 1]
        );
    }

    #[test]
    fn total_nt_closes_a_batch_when_the_running_length_reaches_the_chunk() {
        let reads: Vec<SweepRead> = (0..8).map(|i| r(i, 0, 100, 1.0)).collect();
        // total 800, /4 = 200, +1 = 201; so batches close after 3 reads (300)
        let b = batch_list(&reads, 4, BatchType::TotalNt);
        assert_eq!(b.iter().map(|x| x.len()).collect::<Vec<_>>(), vec![3, 3, 2]);
    }

    /// The final yield is UNCONDITIONAL, so a run whose last chunk lands exactly
    /// on the threshold gets an EMPTY last batch. isONclust guards this with
    /// `if batch:` and NGSpeciesID does not.
    ///
    /// **Neither committed corpus reaches it.** Swept every (corpus, --t,
    /// --batch_type) combination in the case matrix: not one produces an empty
    /// batch, so the goldens cannot tell the guarded version from the unguarded
    /// one. That is exactly why this test exists -- a change nothing exercises
    /// is a change nobody has checked.
    ///
    /// The expected values were measured by calling the reference's own
    /// `parallelize.batch_list`, not derived here:
    ///
    /// ```text
    /// n=2  L=10 cores=2 -> 2 batches, sizes [2, 0]   <-- empty
    /// n=6  L=10 cores=3 -> 3 batches, sizes [3, 3, 0]
    /// n=4  L=10 cores=2 -> 2 batches, sizes [3, 1]   (no empty)
    /// ```
    ///
    /// It matters because the batch count drives the merge-iteration count,
    /// which decides how many numbered `<n>/` directories a run writes -- and
    /// those are recorded output.
    #[test]
    fn the_final_batch_is_yielded_even_when_empty() {
        for (n, len, cores, want) in [
            (2usize, 10usize, 2usize, vec![2usize, 0]),
            (6, 10, 3, vec![3, 3, 0]),
            // ...and the neighbouring shapes that do NOT produce one, so a
            // change that starts emitting empties everywhere also fails.
            (4, 10, 2, vec![3, 1]),
            (3, 10, 3, vec![2, 1]),
            (10, 10, 5, vec![3, 3, 3, 1]),
        ] {
            let reads: Vec<SweepRead> = (0..n).map(|i| r(i, 0, len, 1.0)).collect();
            let b = batch_list(&reads, cores, BatchType::TotalNt);
            assert_eq!(
                b.iter().map(|x| x.len()).collect::<Vec<_>>(),
                want,
                "n={n} len={len} cores={cores}"
            );
        }
    }

    #[test]
    fn merge_consecutive_closes_on_a_rising_threshold() {
        // score order, batch indices crossing 2 then 4
        let reads = vec![
            r(0, 1, 10, 9.0),
            r(1, 2, 10, 8.0),
            r(2, 3, 10, 7.0),
            r(3, 4, 10, 6.0),
        ];
        let b = batch_list_merge_consecutive(&reads);
        // ids 0,1 have index <= 2; then 2 opens a new batch (threshold -> 4),
        // and 3 has index 4 <= 4 so it joins it.
        assert_eq!(b.len(), 2);
        assert_eq!(b[0].iter().map(|x| x.id).collect::<Vec<_>>(), vec![0, 1]);
        assert_eq!(b[1].iter().map(|x| x.id).collect::<Vec<_>>(), vec![2, 3]);
    }
}
